Improve esp32 support by being less strict

medlor opened the pull request:
less errors by being less strict

medlor commented:
moment.. there are still issues...

medlor commented:
Fixed

TheTrueAI commented on ZACwire.h:
```diff
@@ -93,6 +93,7 @@ class ZACwire {
 		static void attachISR_ESP32(void*){					//attach ISR in freeRTOS because gcc 5.3+ can't put ISR in IRAM inside of template class
 			gpio_num_t isrPin = (gpio_num_t)pin;
 			gpio_pad_select_gpio(isrPin);
+			gpio_set_direction((gpio_num_t)isrPin, GPIO_MODE_INPUT);
```
The GPIO was already set as input in line 32 in begin(), or is there in issue with `pinMode(pin, INPUT)`?

medlor replied:
I tested a lot, and had strange behavior in some situations. finally I dont think this line is required.

TheTrueAI commented on ZACwire.h:
```diff
@@ -122,7 +123,7 @@ class ZACwire {
 					if (bitCounter == 10) microtime += bitWindow>>2;	//convert timestamp at stop bit to normal 0 bit
 					rawData[backUP] <<= 1;
-					rawData[backUP] |= bitWindow > deltaTime + (bitWindow>>1 & -!(rawData[backUP] & 2));	//add 1/2 bitWindow if previous bit was 1 (for normalisation)
+					rawData[backUP] |= bitWindow > deltaTime + ((rawData[backUP] & 2)?0:(bitWindow>>1)); //add 1/2 bitWindow if previous bit was 1 (for normalisation)
```
Good point working on this line. It was far to complicated to read. I would even suggest to split it into two lines like this:
```c++
if (rawData[backUP] & 2) deltaTime += bitWindow >> 1;   //add 1/2 bitWindow if previous bit was 1
rawData[backUP] |= bitWindow > deltaTime;
```

medlor replied:
sounds fine. I will change both points in this PR.

medlor replied:
nevertheless..
extensive testing have shown that on esp32 strange flapping on the temperatures occur non-deterministically (at least I cannot find a common cause). They might happen anytime and seem to be clustered around 1-3 minutes. The wrong readings of the temperatures are most often "shifted" for a couple 1/10th of degrees. Rarely up to 3.2 degrees (at which I can also mostly also see one 222 error). 
Example:
 temp sensor inaccuracy: thirdlastTemperature=22.90, secondlatestTemperature=22.50, latestTemperature=22.90

Non-optimal bitwindow worsen the situation dramatically. Therefore I propose to merge the "auto-tuning bitwindow" PR which prevents the situation already after the first reading. Also my code previously only called getTemp() every second, so that the previous "auto-tuning" took very long.

medlor replied:
> ```c++
> if (rawData[backUP] & 2) deltaTime += bitWindow >> 1;   //add 1/2 bitWindow if previous bit was 1
> rawData[backUP] |= bitWindow > deltaTime;
> ```
Das geht so nicht, da dann deltaTime mehrfach erhöht wird und die Werte falsch sind. Davor wurde es nur in der Prüf-Condition erhöht.

TheTrueAI replied:
It works to add the half bitWindow to the deltaTime, as I successfully tested in the beta branch

medlor committed:
- Added directMode, which disable a few internal mechanism to detect outliers, disables on most part useBackup functionality and takes care that the same temp value is not returned twice.
- This is useful when the calling program needs to have access to "raw" data to build their own "outlier" detection (recommended on esp32).
- Added auto-tuning of initial bitwindow to quicker tune-in to best values. (This helps when using esp32).
- Added getBitWindow() to help debug stuff.
- Minor one-bit-off fix of timeDelta saved in rawData (?).

medlor committed:
Add ZACWIRE_VERSION macro

medlor committed:
Remove leftover return

TheTrueAI commented on ZACwire.h:
```diff
@@ -114,15 +149,16 @@ class ZACwire {
 				deltaTime = microtime - deltaTime;			//measure time to previous rising edge
 				if (deltaTime >> 10) {					//start bit
-					backUP ^= bitCounter == 20;
+					backUP ^= bitCounter >= 20;     //sometimes additional interrupts are thrown afterwards
```
Good idea!

TheTrueAI replied:
Thinking it through made it not looking like a good idea anymore. If unwanted interrupts happen and the bitCounter is more than 20, the raw temperature is corrupt and should be overwritten as soon as possible. So backUP should stay the same in that case, means and `== 20` makes sense

medlor replied:
You are right: This is a left over from previous tests. But still the code is not ideal with "==" either because then "backUP temp" which got collected far in the past is still returned in getTemp().
The new _directMode will also fix this, but perhaps it would be better to add some more logic to stop reading bits after the 19th bit is read? I have seen random, non-deterministic interrupts every now and then (quite often!) on esp32.

TheTrueAI commented on ZACwire.h:
```diff
@@ -58,29 +74,47 @@ class ZACwire {
-			if (bitCounter != 19) useBackup = true;				//when newer reading is incomplete
+			if (bitCounter != 19) useBackup = true;			    //when newer reading is incomplete
 			uint16_t temp = rawData[backUP^useBackup];			//get rawData from ISR
+			if (_directMode) rawData[backUP^useBackup] = 0;     //dont ever return the same temp twice
```
If getTemp() gets called twice within 100ms, the first output will be the temperature and the second one error 222. The error 222 looks for the user like there's something wrong, but actually everything is fine.
I'm not sure if the directMode in general is really necessary for the average user, I'll think about it

medlor replied:
Yeah that's the point of having the directMode. The average user probably dont need this, but every other application who needs to address such kind of hiccups (on esp32 mainly) will be interested in having the possibility to react in software.
Return 222 on directMode tells the application that no new data is available.

lebuni commented:
I merged your auto-tuning of initial bitWindow into the beta branch. Before that I rewrote the function several times to get rid of the delay and to make `begin()` faster. After trying a lot of different stuff, I ended up with just a few lines:
```c++
	while (pulseIn(_pin, LOW, timeout*2)) yield;		//wait for time without transmission
	uint8_t strobeTime = pulseIn(_pin, LOW);		//check for signal and measure strobeTime of start bit
        [...]
	if (!bitWindow) bitWindow = strobeTime<<1;
```
In testing it worked really fine.

medlor commented:
Checking your beta branch, I can see you put in much effort to rewrite all the stuff. Great!

Can you please add _directMode and _core support again? As well as the ZACWIRE_VERSION define?

By manually adding _directMode I can still see flapping, but that is expected on esp32 due to wrong handling of interrupts by the hardware.

lebuni commented:
As I measured with a logical analyzer, the ISR just uses 0.078% of the ESP32 CPU time and takes around 4µs. So it shouldn't disturb any other application running on the same CPU. Thus, I didn't include that _core pinning feature in the new version.

The _directMode I think is just interesting for hardware malfunction. I've tried now three different TSics with three different ESP32 and no combination of those could reproduce that flapping. I think usually people would rather return the faulty hardware to the dealer than writing some code around it.

Big-Brummer commented:
Lieber Adrian Immer, ich denke Sie sprechen Deutsch.
Leider weiß ich Sie nicht anders zu kontaktieren.
Zunächst recht herzlichen Dank für Ihr tolles zacwire Programm.
Ich möchte dieses verwenden um einen 14-Bit Sensor auszulesen.
Mit Ihrem Programm kann ich erfolgreich aber nur 11 Bit lesen.
Meine Versuche Ihr Programm auf 14 Bit abzuändern scheitern bzw. 
nach wenigen Sekunden stürzt mein Programm ab.
Ich wäre Ihnen Sehr dankbar für einen Tipp wie ich zacwire auf 14-Bit abändern kann.
ganz herzlichen Dank
J. Steger
